File: csharp/1388.pizza-with-3-n-slices.cs
/*
 * @lc app=leetcode id=1388 lang=csharp
 *
 * [1388] Pizza With 3n Slices
 */

using System;
using System.Linq;

// @lc code=start
public class Solution {
    
    private static int MaxSizeSlices(int[] slices, int k) {
        int n = slices.Length;
        int[,] dp = new int[n + 1, k + 1];
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= k; ++j) {
                if (i == 1) {
                    dp[i, j] = slices[0];
                } else {
                    dp[i, j] = Math.Max(dp[i - 1, j], dp[i - 2, j - 1] + slices[i - 1]);
                }
            }
        }
        return dp[n, k];
    }
    
    public int MaxSizeSlices(int[] slices) {
        int n = slices.Length, k = n / 3;
        return Math.Max(MaxSizeSlices(slices.SkipLast(1).ToArray(), k), MaxSizeSlices(slices.Skip(1).ToArray(), k));
    }
}
// @lc code=end

File: java/1388.pizza-with-3-n-slices.java
/*
 * @lc app=leetcode id=1388 lang=java
 *
 * [1388] Pizza With 3n Slices
 */

// @lc code=start
class Solution {
    public int maxSizeSlices(int[] slices) {
        return Math.max(helper(slices, 0), helper(slices, 1));
    }

    private static int helper(int[] slices, int offset) {
        int l = slices.length - 1, n = slices.length / 3;
        int[][] dp = new int[l + 1][n + 1];
        for (int i = 1; i <= l; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (i == 1) {
                    dp[i][j] = slices[offset];
                } else {
                    dp[i][j] = Math.max(
                        slices[offset + i - 1] + dp[i - 2][j - 1],
                        dp[i - 1][j]
                    );
                }
            }
        }
        return dp[l][n];
    }
}
// @lc code=end

